pkg/fieldpath/object/object.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package object
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+
+	v1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+)
+
+// ToPaved tries to convert a runtime.Object into a *Paved via
+// runtime.DefaultUnstructuredConverter if needed. Returns the paved if
+// the conversion is successful along with whether the
+// runtime.DefaultUnstructuredConverter has been employed during the
+// conversion.
+func ToPaved(o runtime.Object) (*fieldpath.Paved, bool, error) {
+	if u, ok := o.(interface{ UnstructuredContent() map[string]interface{} }); ok {
+		return fieldpath.Pave(u.UnstructuredContent()), false, nil
+	}
+
+	oMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+	if err != nil {
+		return nil, false, err
+	}
+	return fieldpath.Pave(oMap), true, nil
+}
+
+// PatchFieldValueToObject applies the value to the "to" object at the given
```

Comment:
I'm still a little uncomfortable with the fact that we're adding what appear to be convenience functions that operate on `runtime.Object` for patching, but not for any other part of the `fieldpath` package.

My concern is not in thinking that `runtime.Object` is the wrong level to operate at, but rather that merging now appears to be a special case that is asymmetrical from the rest of the API. If the caller wants to set or get a value they handle conversion to a `*Paved` and use `SetValue` and `GetValue`. However if the caller wants to patch, they use this special package that operates on `runtime.Object`.

Perhaps this makes sense if you imagine this package would be expanded to handle more than just merging with time?

Reply:
I agree. That's why I had also suggested that we can maybe move these convenience functions to core crossplane. The `Paved.MergeValue` is coherent with the rest of the `fieldpath` package, but we also need some higher level functions that will extract the value to be merged into the destination from a source `runtime.Object`, as you also acknowledge. 

What do you think about moving these convenience functions to `crossplane/crossplane` (I'm calling this as "core crossplane", not sure if it conveys the right meaning)? I'm also fine if they stay under `fieldpath/object` as they act as a bridge between `runtime.Object`s and `Paved`s. Maybe a little less for `object.MergeReplace`, but then even `object.MergeReplace` accepts a `path` parameter.

Reply:
Hrm I can see a case either way. Perhaps we should start with these in core Crossplane - we can move them here if we find our library of functions that operate on objects grows.

Reply:
I've moved the functions in `fieldpath.object` to core Crossplane including `resource.WithMergeOptions` and unexported them. If we need to use them elsewhere, we can consider to export them but for now, no need to export them. Please see the dependent [PR](https://github.com/crossplane/crossplane/pull/2409).